src/structs/vpx_bool_reader.rs

```diff
@@ -65,46 +65,149 @@ impl<R: Read> VPXBoolReader<R> {
+
+        if bit {
+            *tmp_range -= split;
+            *tmp_value -= split;
+        } else {
+            *tmp_range = split;
+        }
+
+        let shift = (*tmp_range).leading_zeros() as i32;
+
+        *tmp_value <<= shift;
+        *tmp_range <<= shift;
+        *tmp_count -= shift;
+
+        #[cfg(feature = "compression_stats")]
+        {
+            self.model_statistics
+                .record_compression_stats(_cmp, 1, i64::from(shift));
+        }
+
+        #[cfg(feature = "detailed_tracing")]
+        {
+            self.hash.hash(branch.get_u64());
+            self.hash.hash(tmp_value);
+            self.hash.hash(tmp_count);
+            self.hash.hash(tmp_range);
+
+            //if hash == 0x88f9c945
+            {
+                let hash = self.hash.get();
+
+                print!("({0}:{1:x})", bit as u8, hash);
+                if hash % 8 == 0 {
+                    println!();
+                }
+            }
+        }
+
+        bit
+    }
+
     #[inline(never)]
     pub fn get_grid<const A: usize>(
         &mut self,
         branches: &mut [Branch; A],
-        cmp: ModelComponent,
+        _cmp: ModelComponent,
     ) -> Result<usize> {
         // check if A is a power of 2
         assert!((A & (A - 1)) == 0);
 
+        let mut tmp_value = self.value;
+        let mut tmp_range = self.range;
+        let mut tmp_count = self.count;
+
         let mut decoded_so_far = 1;
 
-        for _index in 0..A.ilog2() {
-            let cur_bit = self.get(&mut branches[decoded_so_far], cmp)? as usize;
+        for index in 0..A.ilog2() {
+            // we can read only each 8-th iteration: minimum 56 bits are in value after vpx_reader_fill,
+            // and one get_bit consumes at most 7 bits (with range coming from >127 to 1)
+            if index & 7 == 0 {
+                Self::vpx_reader_fill(&mut tmp_value, &mut tmp_count, &mut self.upstream_reader)?;
+            }
+
+            let cur_bit = Self::get_bit(
+                &mut branches[decoded_so_far],
+                &mut tmp_value,
+                &mut tmp_range,
+                &mut tmp_count,
+            ) as usize;
             decoded_so_far <<= 1;
             decoded_so_far |= cur_bit;
         }
 
         // remove set leading bit
         let value = decoded_so_far ^ A;
 
+        self.value = tmp_value;
+        self.range = tmp_range;
+        self.count = tmp_count;
+
         Ok(value)
     }
 
     #[inline(never)]
     pub fn get_unary_encoded<const A: usize>(
         &mut self,
         branches: &mut [Branch; A],
-        cmp: ModelComponent,
+        _cmp: ModelComponent,
     ) -> Result<usize> {
+        let mut tmp_value = self.value;
+        let mut tmp_range = self.range;
+        let mut tmp_count = self.count;
+
         let mut value = 0;
 
         while value != A {
-            let cur_bit = self.get(&mut branches[value], cmp)?;
+            if value & 7 == 0 {
```

Comment:
is this really better than tmp_count < 0? Seems like in both cases the branch predictor would realize that it doesn't happen very much

Reply:
I have tried many options for each of three places, and this one is the fastest. I suppose here and in the other case of the same condition no branch is really used as loop can be fully unrolled, but I have not checked the assembly.

Reply:
Looks great... can you add a comment in both places and I can approve this

Reply:
Done.